docs/content/guides/dagster/migrating-to-pythonic-resources-and-config.mdx

```diff
@@ -0,0 +1,295 @@
+from contextlib import contextmanager
+
+from dagster import ConfigurableResource, asset
+
+class FancyDbResource(ConfigurableResource):
+    conn_string: str
+
+    @contextmanager
+    def get_client(self) -> Iterator[FancyDbClient]:
+        try:
+            some_expensive_setup()
+            yield FancyDbClient(self.conn_string)
+        finally:
+            some_expensive_teardown()
+
+@asset
+def asset_one(fancy_db: FancyDbResource) -> None:
+    with fancy_db.get_client() as client:
+        client.execute_query("SELECT * FROM foo")
+```
+
+## Passing a resource directly while having the underlying client on the context
+
+This change to have the user manage accessing objects out of a resource rather than having the @resource factory function do it presents a problem with dealing with an at-scale migration: the old code will expect the client on the `context` object while the new code will except the enclosing resource object when it is accessed as a parameter.
+
+```python file=/guides/dagster/migrating_to_python_resources_and_config/migrating_resources.py  startafter=begin_broken_unmigrated_code endbefore=end_broken_unmigrated_code dedent=4
+@asset(required_resource_keys={"fancy_db"})
+def existing_asset(context) -> None:
+    # This code is now broken because the resource is no longer a FancyDbClient
+    # but it is a FancyDbResource.
+    context.resources.fancy_db.execute_query("SELECT * FROM foo")
+```
+
+In effect, what we want is for the underlying client to live on the context in old code, but have the new resource passed to the asset in new code. To support this we have framework support, in the form of `IAttachDifferentObjectToOpContext`. By implementing this interface you instructing the framework to place a different object on the context object. The idea here is that you implement this while you migrating your code, so that un-migrated and migrated code and co-exist.
+
+```python file=/guides/dagster/migrating_to_python_resources_and_config/migrating_resources.py  startafter=begin_new_third_party_resource_with_interface endbefore=end_new_third_party_resource_with_interface dedent=4    return FancyDbClient(self.conn_string)
+from dagster import ConfigurableResource, IAttachDifferentObjectToOpContext, asset
+
+class FancyDbResource(ConfigurableResource, IAttachDifferentObjectToOpContext):
+    conn_string: str
+
+    def get_object_to_set_on_execution_context(self) -> FancyDbClient:
+        return self.get_client()
+
+    def get_client(self) -> FancyDbClient:
+        return FancyDbClient(self.conn_string)
+
+@asset
+def new_asset(fancy_db: FancyDbResource) -> None:
+    client = fancy_db.get_client()
+    client.execute_query("SELECT * FROM foo")
+
+@asset(required_resource_keys={"fancy_db"})
+def existing_asset(context) -> None:
+    # This code now works because context.resources.fancy_db is now a FancyDbClient
+    context.resources.fancy_db.execute_query("SELECT * FROM foo")
+```
```

Comment:
short call-to-action referring to slack etc might be good - may also be good to include a brief section about migrating use of 3rd party resources (it's simpler than migrating your own, but some users might be curious about that specific case)